File: src/main/java/org/embulk/output/orc/OrcOutputPluginHelper.java
package org.embulk.output.orc;

import com.amazonaws.auth.profile.ProfileCredentialsProvider;
import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.AmazonS3Client;
import com.amazonaws.services.s3.model.DeleteObjectRequest;
import com.google.common.base.Joiner;
import com.google.common.base.Splitter;
import com.google.common.base.Throwables;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Arrays;
import java.util.List;

class OrcOutputPluginHelper
{
    protected OrcOutputPluginHelper()
    {
        throw new UnsupportedOperationException();
    }

    static void removeOldFile(String fpath, PluginTask task)
    {
        // NOTE: Delete a file if local-filesystem, not HDFS or S3.
        String schema = getSchema(fpath);
        if (isDeleteTarget(schema)) {
            switch (schema) {
                case "file":
                    try {
                        Files.deleteIfExists(Paths.get(fpath));
                    }
                    catch (IOException e) {
                        Throwables.propagate(e);
                    }
                    break;
                case "s3":
                case "s3n":
                case "s3a":
                    AmazonS3URILikeObject s3Url = parseS3Url(fpath);
                    AmazonS3 s3client = new AmazonS3Client(new ProfileCredentialsProvider());
                    if (task.getEndpoint().isPresent()) {
                        s3client.setEndpoint(task.getEndpoint().get());
                    }
                    s3client.deleteObject(new DeleteObjectRequest(s3Url.getBucket(), s3Url.getKey()));
                default:
                    // TODO: Unsupported
            }
        }
    }

    public static boolean isDeleteTarget(String schema)
    {
        switch (schema) {
            case "file":
                return true;
            case "s3":
            case "s3a":
            case "s3n":
                return true;
            default:
                return false;
        }
    }

    static String getSchema(String fpath)
    {
        String schema = Splitter.on("://")
                .splitToList(fpath).get(0);
        if (schema.equals("s3a") || schema.equals("s3n") || schema.equals("s3")) {
            return schema;
        }
        else {
            Path path = Paths.get(fpath);
            return path.getFileSystem().provider().getScheme();
        }
    }

    static AmazonS3URILikeObject parseS3Url(String s3url)
    {
        List<String> parts = Arrays.asList(
                s3url.split("(://|/)"));
        String bucket = parts.get(1);
        String key = Joiner.on("/").join(parts.subList(2, parts.size()));
        return new AmazonS3URILikeObject(bucket, key);
    }

    static class AmazonS3URILikeObject
    {
        String bucket;
        String key;

        public AmazonS3URILikeObject(String bucket, String key)
        {
            this.bucket = bucket;
            this.key = key;
        }

        public String getBucket()
        {
            return bucket;
        }

        public String getKey()
        {
            return key;
        }
    }
}

File: src/main/scala/org/embulk/output/orc/OrcOutputPluginHelper.scala
package org.embulk.output.orc

import java.io.IOException
import java.nio.file.{Files, Paths}

import com.amazonaws.auth.profile.ProfileCredentialsProvider
import com.amazonaws.services.s3.AmazonS3Client
import com.amazonaws.services.s3.model.DeleteObjectRequest
import com.google.common.base.Throwables

import scala.beans.BeanProperty

object OrcOutputPluginHelper {
  def removeOldFile(fpath: String, task: PluginTask) = {
    // NOTE: Delete a file if local-filesystem, not HDFS or S3.
    val schema = getSchema(fpath)
    if (isDeleteTarget(schema)) schema match {
      case "file" =>
        try Files.deleteIfExists(Paths.get(fpath))
        catch {
          case e: IOException => Throwables.throwIfUnchecked(e)
        }
      case "s3" | "s3n" | "s3a" =>
        val s3Url = parseS3Url(fpath)
        val s3client = new AmazonS3Client(new ProfileCredentialsProvider)
        if (task.getEndpoint.isPresent) s3client.setEndpoint(task.getEndpoint.get)
        s3client.deleteObject(new DeleteObjectRequest(s3Url.bucket, s3Url.key))
      case _ =>
    }
  }

  def isDeleteTarget(schema: String): Boolean = schema match {
    case "file" => true
    case "s3" | "s3n" | "s3a" => true
    case _ => false
  }

  def getSchema(fpath: String) = {
    val schema = fpath.split("://").toList.apply(0)
    schema match {
      case "s3" | "s3a" | "s3n" => schema
      case _ => {
        val path = Paths.get(fpath)
        path.getFileSystem.provider.getScheme
      }
    }
  }

  def parseS3Url(s3url: String) = {
    val parts = s3url.split("(://|/)").toList
    val bucket = parts.apply(1)
    val key = parts.slice(2, parts.size).mkString("/")
    new OrcOutputPluginHelper.AmazonS3URILikeObject(bucket, key)
  }

  case class AmazonS3URILikeObject(@BeanProperty bucket: String, @BeanProperty key: String)
}
